refactor: allow callable in `CompliantExpr.from_column_names` to reuse for `CompliantNamespace.<all|col|exclude>`

FBruzzesi opened the pull request:
## What type of PR is this? (check all applicable)

- [x] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues

Follow-up on [comment](https://github.com/narwhals-dev/narwhals/pull/2122#discussion_r1976370057)

dangotbanned commented on narwhals/_arrow/namespace.py:
```diff
@@ -117,37 +117,29 @@ def __init__(
         from narwhals._arrow.expr import ArrowExpr
 
+        def get_column_names(_: ArrowDataFrame) -> Sequence[str]:
+            return column_names
+
         return ArrowExpr.from_column_names(
-            *column_names, backend_version=self._backend_version, version=self._version
+            get_column_names=get_column_names,
+            function_name="col",
+            backend_version=self._backend_version,
+            version=self._version,
```
I think `get_column_names` is a bad name for this.

I chose that name to mean **exactly** this:

https://github.com/narwhals-dev/narwhals/blob/9d9fde2b43af33746efce9d730c6b97c2681bc51/narwhals/utils.py#L1355-L1356

Where `get_columns` could easily be mistaken for being related to these:

https://github.com/narwhals-dev/narwhals/blob/9d9fde2b43af33746efce9d730c6b97c2681bc51/narwhals/typing.py#L88-L99

---

Maybe its a personal taste thing - but I think something like `evaluate_*` makes a lot more sense because you're performing a transformation

dangotbanned commented on narwhals/_arrow/expr.py:
```diff
@@ -87,12 +87,11 @@ def func(df: ArrowDataFrame) -> list[ArrowSeries]:
     def from_column_names(
         cls: type[Self],
-        *column_names: str,
+        get_column_names: Callable[[ArrowDataFrame], Sequence[str]],
```
Putting (https://github.com/narwhals-dev/narwhals/pull/2134#discussion_r1976732319) aside - how about:

```py
@classmethod
    def from_column_names(
        cls: type[Self],
        get_column_names: Sequence[str],
        ...
```

Then you could avoid needing to create a new function for each backend and just reuse `utils.get_column_names` in `evaluate_output_names=get_column_names,`

- https://github.com/narwhals-dev/narwhals/pull/2102#issuecomment-2685117308

dangotbanned replied:
Ah I saw this covers `col` and `exclude`.

In that case maybe:
```py
get_column_names: Sequence[str] | Callable[[ArrowDataFrame], Sequence[str]]
```

And then add a check for `callable(...)` somewhere?

dangotbanned commented on narwhals/_dask/namespace.py:
```diff
@@ -67,29 +67,27 @@ def func(df: DaskLazyFrame) -> list[dx.Series]:
                 column_name
                 for column_name in df.columns
                 if column_name not in excluded_names
             ]
```
ooh now that I'm thinking about it - isn't this the part we'd want to factor-out?

```py
# nw.utils.py (or a method on something)

def exclude_columns(frame: _StoresColumns, names: Container[str]):
    return [name for name in frame.columns if name not in names]
```

That should work for every backend IIRC?

dangotbanned replied:
You could even do this if you needed to match the single-arg signature?

```py
from functools import partial

fn = partial(exclude_columns, names=excluded_names)
```

dangotbanned replied:
Resolved in (https://github.com/narwhals-dev/narwhals/pull/2134/commits/efd5a227a082adebffd11a36fe3149445edacf3a)

dangotbanned commented:
Apologies for the *stream-of-consciousness* @FBruzzesi 😉

FBruzzesi commented:
> Apologies for the _stream-of-consciousness_ @FBruzzesi 😉

No worries! There are a lot of good points and I realized `all` could be refactored as well! I should have opened this as draft to begin with. 

I am quite happy with it now - I should have addressed all your points except [comment](https://github.com/narwhals-dev/narwhals/pull/2134#discussion_r1976735018) for which I have mixed feelings.

FBruzzesi commented:
Thanks @dangotbanned 🙌🏼
